hal/spi_api.h

```diff
@@ -19,220 +18,305 @@
+    bool clock_polarity; /**< True if the clock's rest state is high (+Vcc). */
+    uint32_t word_length; /**< Length of a symbol in bit. */
+} spi_init_t;
+
+/**
+ * This enumerates the possible result of the spi_init function.
  */
-void spi_init(spi_t *obj, PinName mosi, PinName miso, PinName sclk, PinName ssel);
+typedef enum spi_result_t {
+    SPI_RESULT_OK, /**< Operation successful. */
+    SPI_RESULT_CONFIG_UNSUPPORTED, /**< The required parameters are not supported by the device. */
+    SPI_RESULT_INVALID_PARAM, /**< The given parameter(s) is/are invalid. */
+    SPI_RESULT_ALREADY_INITIALIZED, /**< The requested peripheral/SS pin is already initialized. */
+} spi_result_t;
 
-/** Release a SPI object
+/**
+ * This structure groups all required data to handle a SPI transaction.
  *
- * TODO: spi_free is currently unimplemented
- * This will require reference counting at the C++ level to be safe
+ * - If tx is NULL a fill symbol is sent instead. See spi_init_t.
+ * - If tx is not NULL, the tx buffer must be at least `tx_count` symbols long.
  *
- * Return the pins owned by the SPI object to their reset state
- * Disable the SPI peripheral
- * Disable the SPI clock
- * @param[in] obj The SPI object to deinitialize
- */
-void spi_free(spi_t *obj);
-
-/** Configure the SPI format
+ * - If rx is NULL, all read symbols are dropped.
+ * - If rx is not NULL, the rx buffer must be at least `rx_count` symbols long.
  *
- * Set the number of bits per frame, configure clock polarity and phase, shift order and master/slave mode.
- * The default bit order is MSB.
- * @param[in,out] obj   The SPI object to configure
- * @param[in]     bits  The number of bits per frame
- * @param[in]     mode  The SPI mode (clock polarity, phase, and shift direction)
- * @param[in]     slave Zero for master mode or non-zero for slave mode
+ * A symbol might be bigger than a byte. In such case symbols are read/written following the
+ * platform's endianness.
  */
-void spi_format(spi_t *obj, int bits, int mode, int slave);
+typedef struct spi_transfer_args_t {
+    const uint8_t *tx; /**< A buffer containing the data to be sent. */
+    uint32_t tx_count; /**< The number of symbol to send. */
+    uint8_t *rx; /**< A buffer to store the received data. */
+    uint32_t rx_count; /**< The number of symbol to read. */
+} spi_transfer_args_t;
 
-/** Set the SPI baud rate
- *
- * Actual frequency may differ from the desired frequency due to available dividers and bus clock
- * Configures the SPI peripheral's baud rate
- * @param[in,out] obj The SPI object to configure
- * @param[in]     hz  The baud rate in Hz
- */
-void spi_frequency(spi_t *obj, int hz);
 
-/**@}*/
 /**
- * \defgroup SynchSPI Synchronous SPI Hardware Abstraction Layer
- * @{
+ * This enumerates the possible event types generated by the SPI ASYNC api.
  */
+typedef enum spi_event_type_t {
+    SPI_EVENT_TYPE_ON_DONE, /**< The operation has completed successfully. */
```

Comment:
Are we changing callbacks to `on_done` or `on_abort` - separate callbacks for each event? (previously events were bitmasked and a user could choose multiple).